src/common/transformations/src/transformations/common_optimizations/matmul_split_decomposition.cpp

```diff
@@ -0,0 +1,209 @@
+                        return false;
+                    }
+                } else {
+                    return false;
+                }
+
+                const auto indices_node = as_type_ptr<opset1::Constant>(gather->get_input_node_shared_ptr(1));
+                if (indices_node) {
+                    const auto& indices_val = indices_node->cast_vector<int32_t>();
+                    if (indices_val.size() != 1u) {
+                        return false;
+                    }
+                    if (indices_val[0] < 0 || indices_val[0] >= 3) {
+                        return false;
+                    }
+                    gathers[indices_val[0]] = gather;
+                    fake_quantizes[indices_val[0]] = fq;
+                } else {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
+
+        if (std::any_of(gathers.begin(), gathers.end(), [](const std::shared_ptr<Node> node_ptr) {
+                return !node_ptr || !is_type<ov::op::util::GatherBase>(node_ptr);
+            })) {
+            return false;
+        }
+
+        Output<Node> bias;
+        OutputVector new_weights, new_bias;
+        if (add) {
+            bias = pattern_map.at(bias_pattern);
+        }
+
+        if (!split_weights(weights, new_weights, (add != nullptr) ? &bias : nullptr, new_bias, transpose_b)) {
+            return false;
+        }
+
+        if (new_weights.size() != decompose_num || ((add != nullptr) && new_bias.size() != decompose_num)) {
+            return false;
+        }
+
+        // Heuristics: Split at axis 2, new Gahter should remove it.
+        const auto const_indices = register_new_node(op::v0::Constant::create(element::i32, Shape{4}, {0, 1, 3, 4}));
+        const auto const_axis = register_new_node(op::v0::Constant::create(element::i32, Shape{}, {0}));
+        const auto new_shape = register_new_node<op::v8::Gather>(reshape_input1, const_indices, const_axis);
+        const auto& input = pattern_map.at(input_pattern);
+        for (size_t i = 0; i < decompose_num; i++) {
+            const auto new_mm = register_new_node<op::v0::MatMul>(input, new_weights[i], false, transpose_b);
+            std::shared_ptr<ov::Node> reshape_productor = new_mm;
+            if (add) {
+                reshape_productor = register_new_node<op::v1::Add>(new_mm, new_bias[i]);
+            }
+            const auto new_reshape = register_new_node<op::v1::Reshape>(reshape_productor, new_shape, true);
+            ov::NodeVector from_nodes = {gathers[i], weights.get_node_shared_ptr(), matmul};
+            if (add) {
+                from_nodes.emplace_back(add);
+                from_nodes.emplace_back(pattern_map.at(bias_pattern).get_node_shared_ptr());
+            }
+            if (as_type<opset1::Transpose>(root_node.get()))
+                from_nodes.emplace_back(root_node);
+
+            copy_runtime_info(from_nodes, get_new_nodes());
```

Comment:
should we move "copy_runtime_info" down after "new_transpose" creation?

Reply:
Done